google/cloud/spanner/numeric.h

```diff
@@ -0,0 +1,269 @@
+  std::string&& ToString() && { return std::move(rep_); }
+  ///@}
+
+  /// @name Relational operators
+  ///@{
+  friend bool operator==(Numeric const& a, Numeric const& b) {
+    return a.rep_ == b.rep_;
+  }
+  friend bool operator!=(Numeric const& a, Numeric const& b) {
+    return !(a == b);
+  }
+  ///@}
+
+  /// Outputs string representation of the `Numeric` to the provided stream.
+  friend std::ostream& operator<<(std::ostream& os, Numeric const& n) {
+    return os << n.ToString();
+  }
+
+ private:
+  friend StatusOr<Numeric> internal::MakeNumeric(std::string s);
+  explicit Numeric(std::string rep) : rep_(std::move(rep)) {}
+  std::string rep_;  // a valid and canonical NUMERIC representation
+};
+
+namespace internal {
+
+// Like `std::to_string`, but also supports Abseil 128-bit integers.
+template <typename T>
+std::string ToString(T value) {
+  return std::to_string(value);
+}
+std::string ToString(absl::int128 value);
+std::string ToString(absl::uint128 value);
+
+// Forward declarations.
+Status DataLoss(std::string message);
+StatusOr<Numeric> MakeNumeric(std::string s, int exponent);
+
+}  // namespace internal
+
+/**
+ * Construction from a string, in decimal fixed- or floating-point formats
+ * (where there must be digits either before or after the decimal point):
+ *
+ *    [-+]?[0-9]+(.[0-9]*)?([eE][-+]?[0-9]+)?
+ *    [-+]?.[0-9]+([eE][-+]?[0-9]+)?
+ *
+ * Fails on syntax errors or if the conversion would yield a value outside
+ * the NUMERIC range. If the argument has more than `kFracPrec` digits after
+ * the decimal point it will be rounded, with halfway cases rounding away
+ * from zero.
+ */
+StatusOr<Numeric> MakeNumeric(std::string s);
+
+/**
+ * Construction from a double. Fails on NaN or any argument outside the
+ * NUMERIC value range (which includes infinities). If the argument has more
+ * than `kFracPrec` digits after the decimal point it will be rounded, with
+ * halfway cases rounding away from zero.
+ */
+StatusOr<Numeric> MakeNumeric(double d);
+
+/**
+ * Construction from an integer `i`, scaled by 10^`exponent`. Fails on any
+ * (scaled) argument outside the NUMERIC value range.
+ */
+template <typename T, typename std::enable_if<
+                          std::numeric_limits<T>::is_integer, int>::type = 0>
+StatusOr<Numeric> MakeNumeric(T i, int exponent = 0) {
+  return internal::MakeNumeric(internal::ToString(i), exponent);
+}
+
+/**
+ * Conversion to the closest double value, with possible loss of precision.
+ * Always succeeds (i.e., can never overflow, assuming a double can hold
+ * values up to 10^(kIntPrec+1)).
+ */
+inline double ToDouble(Numeric const& n) {
+  return std::atof(n.ToString().c_str());
+}
+
+/**
+ * Conversion to the nearest integer value, scaled by 10^`exponent`, rounding
+ * halfway cases away from zero. Fails when the destination type cannot hold
+ * that value.
+ *
+ * @par Example
+ *
+ * @code
+ * spanner::Numeric n = spanner::MakeNumeric(123456789, -2).value();
```

Comment:
I like this example. It nicely shows a general overview of how to use this class, including it's non-member functions (still part of the class's interface).

It makes me think that this example (or an _additional_ example) may be nice up in the class's documentation so that readers of the class docs get a quick overview of how to use the class, including the non-member functions. i'd vote for adding an additional example up there.